app/gui2/src/util/ast/index.ts

```diff
@@ -0,0 +1,247 @@
+  const leftOffsetEnd = leftOffsetbegin + node.whitespaceLengthInCodeParsed
+  const end = leftOffsetEnd + node.childrenLengthInCodeParsed
+  return code.substring(leftOffsetEnd, end)
+}
+
+/**
+ * Read span of code reprsented by given Token.
+ *
+ * The Token is assumed to be a part of AST generated from `code`.
+ */
+export function readTokenSpan(token: Token, code: string): string {
+  const begin = token.startInCodeBuffer
+  const end = begin + token.lengthInCodeBuffer
+  return code.substring(begin, end)
+}
+
+/**
+ * Read direct AST children.
+ */
+// TODO[ao] This is a very hacky way of implementing this. The better solution would be to generate
+// such a function from parser types schema. See parser-codegen package.
+export function* childrenAstNodes(obj: unknown): Generator<Tree> {
+  function* astNodeOrChildren(obj: unknown) {
+    if (obj instanceof Tree.AbstractBase) {
+      yield obj as Tree
+    } else {
+      yield* childrenAstNodes(obj)
+    }
+  }
+  if (obj == null) return
+  const maybeIterable = obj as Record<symbol, unknown>
+  const isIterator = typeof maybeIterable[Symbol.iterator] == 'function'
+  if (obj instanceof LazyObject) {
+    const prototype = Object.getPrototypeOf(obj)
+    const descriptors = Object.getOwnPropertyDescriptors(prototype)
+    const objWithProp = obj as unknown as Record<string, unknown>
+
+    for (const prop in descriptors) {
+      if (descriptors[prop]?.get == null) continue
+      const value = objWithProp[prop]
+      yield* astNodeOrChildren(value)
+    }
+  } else if (isIterator) {
+    const iterable = obj as Iterable<unknown>
+    for (const element of iterable) {
+      yield* astNodeOrChildren(element)
+    }
+  }
+}
+
+/** Returns all AST nodes from `root` tree containing given char, starting from the most nested. */
+export function* astContainingChar(charIndex: number, root: Tree): Generator<Tree> {
+  for (const child of childrenAstNodes(root)) {
+    const begin = child.whitespaceStartInCodeParsed + child.whitespaceLengthInCodeParsed
+    const end = begin + child.childrenLengthInCodeParsed
+    if (charIndex >= begin && charIndex < end) {
+      yield* astContainingChar(charIndex, child)
+    }
+  }
+  yield root
+}
+
+if (import.meta.vitest) {
+  const { test, expect } = import.meta.vitest
+
+  // Not working cases commented.
+  test.each([
+    ' foo bar\n',
+    'Data.read\n2 + 2',
+    'Data.read File\n2 + 3',
+    // 'Data.read "File"\n2 + 3',
+    'foo bar=baz',
+    // '2\n + 3\n + 4',
+  ])("AST spans of '%s' are valid", (input) => {
+    const tree = parseEnso(input)
+    const endPos = validateSpans(tree)
+    expect(endPos).toStrictEqual(input.length)
+  })
+
+  test("Reading AST node's code", () => {
+    const code = 'Data.read File\n2 + 3'
+    const ast = parseEnso(code)
+    expect(readAstSpan(ast, code)).toStrictEqual(code)
+    assert(ast.type === Tree.Type.BodyBlock)
+    const lines = Array.from(ast.statements)
```

Comment:
This `lines` variable confused me for abit, I initially interpreted below it as source code lines. Please rename the variable to `statements`.